src/statsmodel.jl

```diff
@@ -106,6 +106,73 @@ function formula end
 formula(m::TableStatisticalModel) = m.mf.f
 formula(m::TableRegressionModel) = m.mf.f
 
+"""
+    termnames(model::StatisticalModel)
+
+Return the names of terms used in the formula of `model`.
+
+This is a convenience method for `termnames(formula(model))`, which returns a
+two-tuple of `termnames` applied to the left and right hand sides of the formula.
+
+For `RegressionModel`s with only continuous predictors, this is the same as
+`(responsename(model), coefnames(model))`.
+
+For models with categorical predictors, the returned names reflect
+the variable name and not the coefficients resulting from
+the choice of contrast coding.
+"""
+termnames(model::StatisticalModel) = termnames(formula(model))
+
+"""
+    termnames(t::FormulaTerm)
+
+Return a two-tuple of `termnames` applied to the left and
+right hand sides of the formula.
+
+!!! note
+    Until `apply_schema` has been called, literal `1` in formulae
+    is interpreted as `ConstantTerm(1)` and will thus appear as `"1"` in the
+    returned term names.
+
+```jldoctest
+julia> termnames(@formula(y ~ 1 + x * y + (1+x|g)))
+("y", ["1", "x", "y", "x & y", "(1 + x) | g"])
+```
+
+Similarly, formulae with an implicit intercept will not have a `"1"`
+in their term names, because the implicit intercept does not exist until
+`apply_schema` is called (and may not exist for certain model contexts).
+
+```jldoctest
+julia> termnames(@formula(y ~ x * y + (1+x|g)))
+("y", ["x", "y", "x & y", "(1 + x) | g"])
+```
+"""
+termnames(t::FormulaTerm) = (termnames(t.lhs), termnames(t.rhs))
+
+"""
+    termnames(term::AbstractTerm)
+
+Return the name of a term.
+
+Return value is either a `String`, an iterable of `String`s or nothing if there
+no associated name (e.g. `termnames(InterceptTerm{false}())`).
+"""
+termnames(::InterceptTerm{H}) where {H} = H ? "(Intercept)" : nothing
```

Comment:
@nalimilan we still need it to return something so that we get don't get an error for `y ~ 0 + x` where there is a valid term names for the whole entity. The problem with nothing comes in the map-reduce step for compound terms.

Reply:
I guess one other option would be to have `termnames` *always* return a String array, which could be empty (e.g. here), contain one element (ContinuousTerm) or multiple elements (CategoricalTerm). We haven't tagged the release yet, so now's the time to do the change :wink:

Reply:
I hadn't realized you had changed it to return `String[]` instead of  `nothing`. There's no inconsistency with `coefnames` anymore then. Let's keep it that way.